### PR TITLE
waitForEvent

### PR DESCRIPTION
**Pull request by telackey**

### CODE REVIEW

File: src/async.js [hunk 1]
```diff
@@ -2,6 +2,10 @@
 // Copyright 2020 DxOS
 //
 
+import assert from 'assert';
+
+const TRUEFN = () => true;
```
Comment (dboreham) [inline src/async.js]: In the code I sent you this had been re-named. Please use that code.

File: src/async.js [hunk 2]
```diff
@@ -117,3 +121,35 @@ export const promiseTimeout = (promise, ms) => {
...
 };
+
+/**
+ * Create a Promise which will resolve when `eventName` is triggered on `eventEmitter`.
+ * If `checkFn` is specified, it must return truthy for the Promise to resolve.
+ * Example:
+ *   const waitFor123 = waitForEvent(recordProcessor, 'finished', record => record.id === '123');
+ *   recordProcessor.startProcessing();
+ *   const record123 = await waitFor123;
+ * @param eventEmitter
```
Comment (dboreham) [inline src/async.js]: Can we have types on the parameters?

File: src/async.test.js [hunk 3]
```diff
@@ -77,3 +77,31 @@ test('promiseTimeout - timed out', async (done) => {
...
+  const emitter = new EventEmitter();
+  const waitFor123 = waitForEvent(emitter, 'done', record => record.id === '123', 50);
+  setTimeout(() => {
+    emitter.emit('done', { id: '123' });
+  }, 75);
+  try {
+    await waitFor123;
+    done.fail('Timeout not triggered.');
+  } catch (error) {
+    expect(error.message).toMatch(/Timed out/);
```
Comment (dboreham) [inline src/async.test.js:109]: Check time out occurred after the specified time?

**Review by dboreham — CHANGES_REQUESTED** [review]

File: src/async.js [hunk 4]
```diff
@@ -117,3 +121,37 @@ export const promiseTimeout = (promise, ms) => {
     });
   });
 };
+
+/**
+ * Returns a Promise which will resolve when `eventName` is triggered on `eventEmitter`.
+ * If `checkFn` is specified, it must return truthy for the Promise to resolve.
+ * Example:
```
Comment (richburdon) [inline src/async.js:130]: Note required since you have good tests (remove things that might get out of sync).

File: src/async.test.js [hunk 5]
```diff
@@ -77,3 +77,37 @@ test('promiseTimeout - timed out', async (done) => {
     done();
   }
 });
+
+test('waitForEvent', async () => {
+  const emitter = new EventEmitter();
+  const waitFor123 = waitForEvent(emitter, 'done', record => record.id === '123');
```
Comment (richburdon) [inline src/async.test.js:83]: const recordProvider? (these are the semantics we want right)?

File: src/async.js [hunk 6]
```diff
@@ -117,3 +121,37 @@ export const promiseTimeout = (promise, ms) => {
...
+ * @param {number} [timeout]
+ * @returns {Promise<*>}
+ */
+export const waitForEvent = (eventEmitter, eventName, checkFn = noop, timeout = 0) => {
+  assert(eventEmitter);
+  assert(eventEmitter.on);
+  assert(eventName);
+  assert(checkFn);
+
+  const [provider, resolver] = trigger();
```
Comment (richburdon) [inline src/async.js:149]: Now that I understand hooks better (I think) I'm wondering to change tigger to:

```
const [value, setValue] = useValue();

setTimeout(() => setValue(100));

await value();
```

Then useValue could take a generic promise and...

```
const [value, setValue] = useValue(onEvent(emitter, 'foo'));

setTimeout(() => emitter.emit('foo'));

await value();
```

Seems more flexible, simpler and composable? (I had been worrying about "trigger").

```
export const onEvent = (eventEmitter, eventName, test) => {
  return new Promise((resolve, reject) => {
    const off = addListener(eventEmitter, eventName, (...args) => {
      const value = test(...args);
      if (value !== undefined) {
        off();
        resolve(value);
      }
    });
  });
}
```

WDYT?

**Review by richburdon — CHANGES_REQUESTED** [review]

File: src/async.js (hunk 6 above)
Comment (dboreham) [reply src/async.js:149]: Definitely nicer.

Comment (telackey) [reply src/async.js:149]: I don't favor this part:

`const value = test(...args);`

I think a test function should just give a pass/fail result, not potentially manipulate the response.

```
      if (test(...args)) {
        off();
        resolve(...args);
      }
```

**Pull request closed without merging**